src/utils/sort.hpp

```diff
@@ -31,6 +31,28 @@
 
 namespace parthenon {
 
+// Returns the upper bound (or the array size if value has not been found)
+// Could/Should be replaced with a Kokkos std version once available (currently schedule
+// for 4.2 release).
+template <class T>
```

Comment:
Since you use `extent_int` this is not an arbitrary array type. Maybe a good opportunity for concepts?
```suggestion
template<class T, class = ENABLEIF(implements<kokkos_view(Data)>::value)>
```

Reply:
I'm probably using it wrong because when I tried adding this I got
```
[build] /home/pgrete/src/parthenon/src/outputs/histogram.cpp: In lambda function:
[build] /home/pgrete/src/parthenon/src/outputs/histogram.cpp:318:34: error: no matching function for call to ‘upper_bound(const parthenon::ParArrayGeneric<Kokkos::View<double*, Kokkos::LayoutRight, Kokkos::HostSpace>, parthenon::empty_state_t, int>&, double&)’
[build]   318 |               x_bin = upper_bound(x_edges, x_val) - 1;
[build]       |                       ~~~~~~~~~~~^~~~~~~~~~~~~~~~
[build] In file included from /home/pgrete/src/parthenon/src/outputs/histogram.cpp:53:
[build] /home/pgrete/src/parthenon/src/utils/sort.hpp:39:28: note: candidate: ‘template<class T, class> int parthenon::upper_bound(const T&, Real)’
[build]    39 | KOKKOS_INLINE_FUNCTION int upper_bound(const T &arr, Real val) {
[build]       |                            ^~~~~~~~~~~
[build] /home/pgrete/src/parthenon/src/utils/sort.hpp:39:28: note:   template argument deduction/substitution failed:
[build] /home/pgrete/src/parthenon/src/utils/sort.hpp:38:20: error: no type named ‘type’ in ‘struct std::enable_if<false, int>’
[build]    38 | template <class T, class = ENABLEIF(implements<kokkos_view(T)>::value)>
[build]       |                    ^~~~~
```

Reply:
Oooh maybe `ParArrayNDGeneric` doesn't satisfy the `kokkos_view` concept? @lroberts36 which concept should @pgrete be using here?

Reply:
Maybe this should really be operating on types that follow the `contiguous_container` concept? As written it appears that this requires a one-dimensional Kokkos view, but couldn't the `arr.extent_int(0)` instances be replaced with `arr.size()` and `arr(m)` be replaced with `arr[m]`? Then this should work for anything that is a contiguous container I believe (which both kokkos views and `ParArrayNDGeneric`s satisfy).